app/services/katello/pulp3/migration_plan.rb

```diff
@@ -105,7 +105,16 @@ def target_link_map
       end
 
       def name_for_content_view(content_view, root_repo)
-        "#{content_view.label}-#{root_repo.label}"
+        name = "#{content_view.label}-#{root_repo.label}"
+
+        if Katello::RootRepository.where(:label => root_repo.label).group(:label).count(:label)[root_repo.label] > 1
+          repo_query = Katello::Repository.joins(:root, :content_view_version => :content_view).
```

Comment:
Now that doesn't work at all for me. This generates the following SQL in my scenario:

`SELECT "katello_repositories".* FROM "katello_repositories" INNER JOIN "katello_root_repositories" ON "katello_root_repositories"."id" = "katello_repositories"."root_id" INNER JOIN "katello_content_view_versions" ON "katello_content_view_versions"."id" = "katello_repositories"."content_view_version_id" INNER JOIN "katello_content_views" ON "katello_content_views"."id" = "katello_content_view_versions"."content_view_id" WHERE (katello_content_views.id != 12) AND "katello_root_repositories"."label" = '2_3_el7_x86_64' AND "katello_content_views"."label" = 'centos7-epel7'`

which is always empty, unless maybe you have the same content view label in different organizations. 12 is the id of the content view 'centos7-epel7', i.e. this is never true for me.

Reply:
I think the original statement was good, you just needed different grouping.

```
Katello::Repository.joins(:root, :content_view_version => :content_view)
.where("#{::Katello::RootRepository.table_name}.label" => root_repo.label,
"#{::Katello::ContentView.table_name}.label" => content_view.label)
```

generates the SQL

```
SELECT "katello_repositories".*
FROM "katello_repositories"
 INNER JOIN "katello_root_repositories" ON "katello_root_repositories"."id" = "katello_repositories"."root_id"
 INNER JOIN "katello_content_view_versions" ON "katello_content_view_versions"."id" = "katello_repositories"."content_view_version_id"
 INNER JOIN "katello_content_views" ON "katello_content_views"."id" = "katello_content_view_versions"."content_view_id"
WHERE "katello_root_repositories"."label" = '2_3_el7_x86_64'
      AND "katello_content_views"."label" = 'centos7-epel7'
```

which correctly contains all repositories which have a root_repo_label of '2_3_el7_x86_64' in any content view with label 'centos7-epel7'. Now to find out whether these are in fact different root_repos which would create the identical names in the migration plan, I think you would have to group by "katello_root_repositories".id and "katello_content_views"."id" resulting in an SQL like this:

```
SELECT "katello_root_repositories".id, "katello_content_views"."id"
FROM "katello_repositories"
 INNER JOIN "katello_root_repositories" ON "katello_root_repositories"."id" = "katello_repositories"."root_id"
 INNER JOIN "katello_content_view_versions" ON "katello_content_view_versions"."id" = "katello_repositories"."content_view_version_id"
 INNER JOIN "katello_content_views" ON "katello_content_views"."id" = "katello_content_view_versions"."content_view_id"
WHERE "katello_root_repositories"."label" = '2_3_el7_x86_64'
   AND "katello_content_views"."label" = 'centos7-epel7'
GROUP BY "katello_root_repositories".id, "katello_content_views"."id";
 id  | id 
-----+----
 138 | 12
 140 | 12
(2 rows)
```
So basically the statement before by with the GROUP BY (and adjusted SELECT expression).

These are all root_repo ids with the identical root_repo label connected to the any content view with the identical content view label. If this query results in more than one row, it's a duplicate which needs the name changed.

To compare: I have multiple repositories labeled "gluster-7" which however do not coexist in the same content view (one is centos8 the other is centos7). Due to that, there is no conflict:
```
SELECT "katello_root_repositories".id, "katello_content_views"."id"
FROM "katello_repositories"
 INNER JOIN "katello_root_repositories" ON "katello_root_repositories"."id" = "katello_repositories"."root_id"
 INNER JOIN "katello_content_view_versions" ON "katello_content_view_versions"."id" = "katello_repositories"."content_view_version_id"
 INNER JOIN "katello_content_views" ON "katello_content_views"."id" = "katello_content_view_versions"."content_view_id"
WHERE "katello_root_repositories"."label" = 'gluster-7' AND "katello_content_views"."label" = 'centos7-epel7'
group by "katello_root_repositories".id, "katello_content_views"."id";
 id | id 
----+----
 82 | 12
(1 row)
```

Now if I only knew ruby and were able to get this into a ruby statement...

Reply:
Another note: I only have a single organization so I don't really know if this can happen or not. As far as I understand the content view label is only unique within one organization, i.e. two organizations could have content views with the identical label. There would also be duplicate identical names if it is possible to have the same root repo (i.e. the same katello_root_repositories.id) in several content views with the same label of different organizations. That's the same issue, just not on the root_repo.label part but on the content_view.label.

name = "#{content_view.label}-#{root_repo.label}"

But I don't know if it is at all possible to share a root_repo between organizations...

Reply:
O.K. By experimenting with the rake console I think I found the all-in-one statement to find out whether the name must be changed or not:

```
Katello::Repository.joins(:root, :content_view_version => :content_view)
.where("#{::Katello::RootRepository.table_name}.label" => root_repo.label,
       "#{::Katello::ContentView.table_name}.label" => content_view.label)
.select('katello_root_repositories.id')
.group('katello_root_repositories.id','katello_content_views.id').count.count
```

If that is > 1 the name must be changed...